lib/datadog/lambda/trace/xray.rb

```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
+
+#
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+#
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019 Datadog, Inc.
+#
+require 'datadog/lambda/trace/constants'
+require 'datadog/lambda/utils/logger'
+require 'securerandom'
+require 'json'
+require 'socket'
+
+module Datadog
+  # Trace contains utilities related to reading/writing trace context from
+  # lambda events/X-Ray
+  module Trace
+    class << self
+      def read_trace_context_from_xray
+        header = ENV[XRAY_ENV_VAR]
+        segment = parse_xray_trace_context_header(header)
+        trace_id = convert_to_apm_trace_id(segment[:xray_trace_id])
+        parent_id = convert_to_apm_parent_id(segment[:xray_parent_id])
+        sample_mode = convert_to_sample_mode(segment[:xray_sample_mode])
+
+        if trace_id.nil? || parent_id.nil? || sample_mode.nil?
+          Datadog::Utils.logger.error("couldn't read xray header #{header}")
+          return nil
+        end
+        {
+          trace_id: trace_id,
+          parent_id: parent_id,
+          sample_mode: sample_mode,
+          source: SOURCE_XRAY
+        }
+      end
+
+      def add_trace_context_to_xray(context)
+        data = generate_xray_metadata_subsegment(context)
+        send_xray_daemon_data(data)
+        Datadog::Utils.logger.debug("sent metadata to xray #{data}")
+      end
+
+      def generate_xray_metadata_subsegment(context)
+        header = ENV[XRAY_ENV_VAR]
+        segment = parse_xray_trace_context_header(header)
+
+        {
```

Comment:
What's this hash used for? If it's meant to reflect all the fields from a trace, I'm concerned it might be a bit brittle as some of these fields may change. (Particularly the use of `sampling-priority` maybe be irrelevant in the near future.)

Reply:
So, this isn't a span from dd-trace, but an x-ray subsegment. Part of making dd-trace work in lambda, is merging [datadog traces with x-ray traces](https://github.com/DataDog/devops/wiki/DD-and-X-Ray-Hybrid-Tracing). We do this by submitting a metadata span to X-Ray which has the dd-trace context, then our x-ray converter uses that to update all the trace ids of the x-ray trace. Previously we were using the X-Ray sdk to do this, but it would refuse to work without being initialised and turning on instrumentation, (which obviously we don't want when dd-trace is enabled). So I stripped out the X-Ray calls and talked to the X-Ray daemon directly using it's http api, which is what this payload is for. We only use the public API's for dd-trace.